homeassistant/components/control4/media_player.py

```diff
@@ -0,0 +1,372 @@
+                if current_data.get(CONTROL4_PAUSED, None):
+                    return MediaPlayerState.PAUSED
+                if current_data.get(CONTROL4_STOPPED, None):
+                    return MediaPlayerState.ON
+            current_source = self._id_to_parent.get(current_source, None)
+        return None
+
+    @property
+    def state(self):
+        """Return whether this room is on or off."""
+
+        if source_state := self._get_current_source_state():
+            return source_state
+
+        if self.coordinator.data[self._idx][CONTROL4_POWER_STATE]:
+            return MediaPlayerState.ON
+
+        if self._is_soft_on:
+            return MediaPlayerState.IDLE
+
+        return MediaPlayerState.OFF
+
+    @property
+    def source(self):
+        """Get the current source."""
+        current_source = self._get_current_playing_device_id()
+        if not current_source or current_source not in self._sources:
+            return None
+        return self._sources[current_source].name
+
+    @property
+    def media_title(self) -> str | None:
+        """Get the Media Title."""
+        current_source = self._get_current_playing_device_id()
+        if not current_source or current_source not in self._sources:
+            return None
+        return self._sources[current_source].name
+
+    @property
+    def media_content_type(self):
+        """Get current content type if available."""
+        current_source = self._get_current_playing_device_id()
+        if not current_source:
+            return None
+        return MediaType.MUSIC
+
+    async def async_media_play_pause(self):
+        """
+        If possible, toggle the current play/pause state.
+
+        Not every source supports play/pause.
+        Unfortunately MediaPlayer capabilities are not dynamic,
+        so we must determine if play/pause is supported here
+        """
+        if self._get_current_source_state():
+            await super().async_media_play_pause()
+
+    @property
+    def source_list(self) -> list[str]:
+        """Get the available source."""
+        return [x.name for x in self._sources.values()]
+
+    @property
+    def volume_level(self):
+        """Get the volume level."""
+        return self.coordinator.data[self._idx][CONTROL4_VOLUME_STATE] / 100
+
+    @property
+    def is_volume_muted(self):
+        """Check if the volume is muted."""
+        return bool(self.coordinator.data[self._idx][CONTROL4_MUTED_STATE])
+
+    async def async_select_source(self, source):
+        """Select a new source."""
+        for avail_source in self._sources.values():
+            if avail_source.name == source:
+                audio_only = _SourceType.VIDEO not in avail_source.source_type
+                await self._create_api_object().setSource(
+                    avail_source.id, audio_only=audio_only
+                )
+                break
+
+        await self.coordinator.async_request_refresh()
+
+    def turn_on(self):
```

Comment:
There is a `async_turn_off` function that actually calls a Control4 API call to turn off the room - is it not possible to tell Control4 to turn the room on?

Reply:
No, the rooms are always "on" in a sense. They become powered when we select a source for the room thus we have this dummy turn on. The speakers can be turned "off" in the traditional sense via the API call.